### ISSUE TITLE
Make adding additional mutations more robust

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/msprime#1440]
#1434 updated the API for sim_mutations when dealing with adding to an existing set of mutations. The current system is pretty unsatisfactory though, because there's no guarantee that an error won't occur, so anyone using this will have to be prepared to wrap it in a try-except and discard any simulations that fail. This isn't very user-friendly.

The problem is that we generate silent mutations (.e.g A -> A) with non-zero probability, and these are not allowed in tskit.

So, we'd like to do better. I'll copy some comments over from #1434.

**Comment by jeromekelleher** [comment]
https://github.com/tskit-dev/msprime/pull/1434#issuecomment-769849028

We definitely thought about this. I think we left it like this because it's not reasonable to do it 'right' in general, so if a user ran into this problem then they got what they asked for and would have to sort it out themselves.

This seems a little unfair - it seems like someone ought to be able to add more parent-independent mutations (eg from JC69) on top of old ones. This would be trivial if we simulated and allowed non-state-changing mutations: we'd get JC69 by throwing down a Poisson process of ACGT on the tree, and so doing it twice with rate 1 is the same as doing it once with rate 2. However, what we do is equivalent to that but followed up by erasing non-state-changing mutations; so if we threw down mutations with rate 1 and got this on a branch:

   A --------> A ------> C
and then erased the derived A:

   A -------------------> C
and then threw down more mutations with rate 1:

   A --> G ------------> C
this would be incorrect (ie we'd be under-simulating mutations), because we should have

   A --> G --> A -----> C
So, that rules out "just put down more JC69 and erase mutations with state equal to their parent".

But, rejection sampling? It is true that if you wanted to add more mutations to

   A -------------------> C
with JC69, then you should (a) place mutations

   A ------> ? --------> C
and then (b) choose an allele, rejection sampling until you get a G or a T:

   A ------> G --------> C
However, this isn't always possible: in a binary alleles model, you just can't insert another (single) state change between two alleles: 0 ------> ? -------> 1.

So, I think the previous decision about rejection sampling was fine: we'd just need to let the user do it themselves. It wouldn't be hard to go back and fix up the tables for JC69, for instance. (And if we were to do rejection sampling, this would probably be the most efficient way to do it, since usually the situation will be rare.)

But, you're right, we should alert the user. We could easily throw an error or warning here when populating the tables, or here when generating mutations. How about an error (warning?) that says "Mutating on top of existing mutations resulted in a derived state is equal to ancestral state: the resulting tree sequence will not be valid. If you meant to do this, please file an issue." We could then provide the "go back and rejection sample alleles to fix up the tables" method if someone does need it.

**Comment by jeromekelleher** [comment]
https://github.com/tskit-dev/msprime/pull/1434#issuecomment-769945674 @petrelharp 

Sounds good, with a few caveats:

we can only check if everything is OK with mutation matrix models
and so we'll still need the add_ancestral option, to e.g. add SLiM mutations above existing ones
this isn't at the top of my list of things to do
OH WAIT: problem. Suppose we have mutations on each of three descendant branches

derived_state parent
A  -1
C   0
G   0
T    0
and then add a mutation to the parent branch:

derived_state parent
A  -1
?   0
C   1
G   1
T    1
... there's no legal choice for ?. The right way to do this is pretty involved: we'd have to also possibly add a reverse mutation (i.e., ?=T but then put in a T->A mutation later) and/or remove the later, pre-existing mutation.

**Comment by jeromekelleher** [comment]
> This would be trivial if we simulated and allowed non-state-changing mutations: we'd get JC69 by throwing down a Poisson process of ACGT on the tree, and so doing it twice with rate 1 is the same as doing it once with rate 2.

I wonder if we can simplify this a bit for the JC69 case (or any model with a 0 diagonal in the transition matrix). Then, aren't silent transitions disallowed anyway?

I don't think it would be terrible if we just allowed ancestral mutations to be added for this case - I think it would cover most use cases (particularly if SLiM simulations were nucleotidified).

**Comment by petrelharp** [comment]
> I wonder if we can simplify this a bit for the JC69 case (or any model with a 0 diagonal in the transition matrix). Then, aren't silent transitions disallowed anyway?

I'm not sure what you're suggesting?

If we allowed silent mutations, then JC69 would be parent-independent, and hence easy: we could just throw down more mutations, and be done. But since we don't allow silent mutations, in a sense *no* model can be parent-independent (although we do simulate exactly from the distribution of state-changing mutations induced by the parent-independent version of JC69, for instance).  And, as my OH WAIT example above shows, it's not even obvious how to add new mutations in a legal way (ie, that doesn't produce silent mutations).

. . .

Ok, I've had a think about the right way to do this. Suppose that we have a model with mutation matrix `M` that can be made parent-independent by introducing silent mutations with probability `s`, i.e., if `U` is defined by
```
  U = (1 - s) M + s I
```
then all the rows of `U` are the same. (This is generalizable, but this applies to JC69 at least.)  Now, mutating using matrix `M` and rate `mu` is the *same* as mutating with matrix `U` and rate `mu / (1 -s)` (since only a fraction `1 - s` of mutations actually change the state). I claim that if we want to mutate some more, with rate `nu`, then what we need to do is to mutate at rate `nu / (1 - alpha)` with matrix
```
 V = (1 - alpha) M + alpha I
```
where `alpha = 1 - 1/(1 + (1 + mu / nu)/(1 - s)) = (1 + mu/nu) / (2 - s - mu/nu)`. (Here `alpha` is the solution to:
```
(mu + nu) U / (1 - s) = mu M + (nu / (1 - alpha)) * ( (1 - alpha) M + alpha I ),
```
... and I may have gotten the solution to this wrong.)  What's going on here is that to be correct, we have to simulate the "extra" silent mutations that have been "forgotten". Happily, we can add these back in, thanks to the magic of Poisson processes.

Notes:
-  this might actually "remove" some existing mutations (i.e., make them silent), but that's actually correct (as the example above shows).
- doing this correctly requires knowing the previous mutation rate (`mu`). So, it's not just a simple call to `sim_mutations( rate = nu, add=TRUE)`; we'd need a `previous_rate_that_were_adding_over=mu` argument as well.

Options:
1. make it so that when adding more mutations, existing mutations might be removed through the process of removing silent mutations; document the math above and let users do it themselves.
2. provide the `previous_rate_that_were_adding_over` argument and implement the math above (would still need to remove existing ones that became silent).
3. tell users to please only add mutations with infinite-sites mutations, for the love of god.
4. allow silent mutations.

But, when will people need to add mutations? Most importantly, if they have a simulation with only selected mutations and they want to add neutral ones. This is why the SLiM mutation model is infinite-alleles, so that this works. (This is also why it's important that the SLiM mutation model "stacks" mutations, so that if a neutral mutation gets added on top of a selected one, the selected-mutation-genotypes are unchanged; what should get written to a VCF is not clear.)  Adding neutral mutations afterwards is an approximation with finite sites and non-stacking mutations, anyhow; which suggests another option:

5. provide a "don't mutate already-mutated sites" argument that just doesn't do anything if it picks a site with a kept mutation on it (I'd say "rejection sample until you get a good site" but this could be impossible).

This wouldn't be right, but once we're talking about non-stacking mutations under selection, it'd be no worse than what we've got already. But I feel like this should be thought about a bit more.

Note that option (3) is the correct way to do things for the SLiM mutation-stacking model.

Beyond that, it's hard to know what to do, because we need to think about a coherent model of selection without mutation stacking, and we don't really have one. Maybe the most natural one is this: suppose that each site in the genome has probability p of being under selection, so that any nucleotide change at such sites will have a selection coefficient. However, we don't at present have a way to simulate only-selected-mutations under this model: we can simulate selected mutations at rate p * mu, but this under-simulates the rate of double mutations at selected sites, because selected mutations at unmutated sites should happen at rate p * mu (since that's rate mu, with probability p of falling at a selected mutation); but then mutations at already-mutated sites should actually happen at rate mu (not the lower rate p * mu).  I think all this is OK (and, SLiM's mutation stacking model as well) since there's a number of changes that can happen to the genome that are not SNPs (methylation, transposons, duplications...), so we're operating at a higher level of abstraction than actually sequence-to-function.

**Comment by jeromekelleher** [comment]
Thanks @petrelharp. Well, this is all quite complicated. So, for msprime 1 (which I want to feature freeze in two weeks) I think we can either:

1. Remove the add_ancestral option entirely so users can't add mutations older than any existing
2. Keep things as they are, and add a custom exception type (SilentMutationError or something) and document that users must expect these to happen and rejection sample them out
3. Do the fancy math you have here, and also provide the APIs for any extra information needed
4. Make tskit support silent mutations.

Option 4 is looking like the most attractive to me at the moment. There's no fundamental reason we can't have silent mutations in tskit, it just seemed like something that made sense to rule out. However, if it means that we can't simulate multiple mutational processes without jumping through major hoops, then that seems like a perfectly good reason for allowing silent mutations. 

I can't see allowing silent mutations breaking any code - that's the nice thing about defaulting to overly strict. Unless someone is expressly depending on detecting the current errors, then there's no way we can break their code by not erroring any more.

**Comment by petrelharp** [comment]
Option (1) is no good, because that'd mean we can't use msprime to add neutral mutations to a SLiM simulation with only selected sites. And, I'd maybe rephrase (2) as "Keep things as they are, strongly encourage users to only use infinite-alleles models, and add a custom exception type...".

But, I agree, option (4) is very attractive. The only caveat is that if we go with (4) we'd also want to change JC69 to insert silent mutations, either always or optionally, which will be confusing to people.  (I guess this is an argument to have it be optional, and say that if you are going to be layering on more mutations then you need to enable them. Or deal with the inconsistency, which would be *quite* small.)  It's a little hard to work out what should happen because we don't actually have a use case where people are wanting to layer on mutations not with an infinite alleles model.

**Comment by jeromekelleher** [comment]
OK, let's go with 4.

I've been thinking about this, and suppose you wanted to have a mixture of mutation models. Say, you wanted JC69 at rate r1, and some other different matrix mutation models m2, m3, at rate r2, r3,.... Would it be valid to do this as 

```python
ts = sim_mutations(sim_mutations(sim_mutations(ts, model=m3),) model=m2), model="JC69")
```

If so, would it be robust to changes in ordering, etc?

If this is true and we have some good mathematical properties from keeping the silent mutations, I think this is a strong argument for keeping the silent mutations *in* JC69 (and all the other models). They'll be rare in most cases and harmless anyway. If someone wants to remove the silent mutations, they can call ``ts.canonicalise()`` (which should remove them?).

**Comment by petrelharp** [comment]
This would be valid, and independent of ordering, as long as they are all parent-independent: you can think of a parent-independent model as just throwing down a Poisson process of mutations, where the mutation says "now change to an A" (or whatever); by additivity of Poisson processes, it's all good.

If some of the models are not parent-independent, then the order *would* matter... but for most purposes, it wouldn't matter in practice - the proportion of mutations that it would affect is only one-half of mean nucleotide diversity (pi) anyhow. We can put the appropriate warning in the docs about this.

>  If someone wants to remove the silent mutations, they can call ts.canonicalise() (which should remove them?).

Heh, sure, why not (although optionally?).

**Comment by petrelharp** [comment]
I won't be able to jump in to help on this right away, FYI.

**Comment by jeromekelleher** [comment]
OK, I'll create some tickets.

Does this mean that we can remove the ``add_ancestral`` parameter, since we expect the output to be "right" unless someone is doing something pretty obscure?

**Comment by jeromekelleher** [comment]
Should be able to close this after we incorporate tskit 0.99.11 (#1514), we remove the current checks for silent mutations and write some statistical tests to ensure we do the right thing when (say) doing rate=x twice compared with rate=2x once, etc.

Do you agree with this @petrelharp?

**Comment by petrelharp** [comment]
Sorry, I missed this.

> Does this mean that we can remove the add_ancestral parameter, since we expect the output to be "right" unless someone is doing something pretty obscure?

That'd be nice! Let's see:
- the upside would be that users adding more mutations to a forwards sim with selected mutations wouldn't have to think about "add_ancestral"
- the downside is that users who mutate more than once with non-parent-independent models would be less likely to be aware that this is statistically wrong.

So, we'd be reducing cognitive burden significantly on maybe a common use case, in exchange for avoiding some very, very slight statistical wrongness. It sounds like a great trade to me: not worrying about the precise mutation model in those rare cases of multiple hits is totally the right thing to do, I think.

I guess that also `add_ancestral` helps avoid bugs where someone means to mutate only the more recent part of the tree but forgets to specify an `end_time`. Or means to only mutate a segment of the genome. But this also seems pretty rare.
